### DISCUSSION TITLE
Download universal apk after publication by upload_to_play_store

### DISCUSSION THREAD

**Discussion by art-shen** [fastlane/fastlane#20370] [Ideas]
Universal APK signed by Google Play is very important in many cases:
- CI artifact publication;
- Publication of signed APK to another store (e.g. Amazon, Huawei);
- Any other check/action with actual release APK.

Currently we can obtain signed universal APK only manually from the Google Play Console as fastlane has no support for such action.
Please provide a way to download it after publication. As I can see it's possible via API.

**Comment by qpre** [comment]
Hey @amal ! have you found a workaround for this ?

**Reply by art-shen** [reply]
@qpre, not yet, although it looks possible.  I'll try to do the PoC, but probably as a Gradle task.

**Comment by ersen-lw** [comment]
+1 I have the same use case. Would be amazing if fastlane could provide this functionality. 

Below is some of my notes on the subject. I have not implemented it but will try at a later point in the `fastfile`. 

There is a Play Publisher API to download generated APK but **it only works for app bundle distributions only.** 

You first call the `list` method (https://developers.google.com/android-publisher/api-ref/rest/v3/generatedapks/list) passing in the package name of the app and the version code in question. 

The response is an array of `GeneratedApksPerSigningKey` objects. 

Each object contains different APKs since app bundles can generate a number of different APKs. 

The  object you want for distribution to other stores is the `generatedUniversalApk` object. This object contains one field called `downloadId` . 

This field is then supplied to the `download` method (https://developers.google.com/android-publisher/api-ref/rest/v3/generatedapks/download) . This method takes in the package name, version code and downloadId. This will download the APK file.

**Comment by cbjerg** [comment]
This is exactly what  I am looking for. Have you come any further?

**Comment by zhgchgli0718** [comment]
I wrote a simple extension to the Fastlane Action that calls the Google API to download the uploaded APK. The code is available here: https://gist.github.com/zhgchgli0718/42f6326225e393e20616ce15f3390c88.
